File: 39-rope/rust/src/main.rs
use serde::Deserialize;
use std::io::{self, Read};

#[derive(Deserialize)]
struct Input {
    parts: Vec<String>,
    queries: Vec<[usize; 2]>,
}

fn solve(parts: &[String], queries: &[[usize; 2]]) -> String {
    // TODO: build a rope from parts, answer [lo, hi) substring queries
    // Return extracted substrings joined by "|"
    let _ = (parts, queries);
    todo\!()
}

fn main() {
    let mut buf = String::new();
    io::stdin().read_to_string(&mut buf).unwrap();
    let inp: Input = serde_json::from_str(&buf).unwrap();
    println\!("{}", solve(&inp.parts, &inp.queries));
}

File: 39-rope/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type input struct {
	Parts   []string `json:"parts"`
	Queries [][]int  `json:"queries"`
}

func solve(parts []string, queries [][]int) string {
	// TODO: build a rope from parts, answer [lo, hi) substring queries
	// Return extracted substrings joined by "|"
	return ""
}

func main() {
	var in input
	if err := json.NewDecoder(os.Stdin).Decode(&in); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	fmt.Println(solve(in.Parts, in.Queries))
}
